File: ai-data-dictionary/apps/backend/src/schemas/lineage.py
"""Pydantic schemas for Lineage API."""

from typing import Optional

from pydantic import BaseModel, ConfigDict


class LineageResponse(BaseModel):
    """Schema for lineage edge response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    upstream_table_id: int
    downstream_table_id: int
    relationship_type: str
    description: Optional[str] = None
    column_mapping: Optional[dict] = None
    confidence_score: Optional[float] = None


class LineageNodeResponse(BaseModel):
    """Schema for lineage graph node."""

    id: int
    name: str
    schema_name: str
    table_type: str
    level: int  # Depth in lineage graph


class LineageEdgeResponse(BaseModel):
    """Schema for lineage graph edge."""

    source: int  # upstream_table_id
    target: int  # downstream_table_id
    relationship_type: str
    label: Optional[str] = None
    column_mapping: Optional[list[dict]] = None  # e.g. [{"referenced_column": "id", "referencing_column": "customer_id"}]
    cardinality: Optional[str] = None  # one_to_one, one_to_many, many_to_many (via join table)
    is_join_table: Optional[bool] = None


class LineageGraphResponse(BaseModel):
    """Schema for complete lineage graph."""

    nodes: list[LineageNodeResponse]
    edges: list[LineageEdgeResponse]
    root_table_id: int


class LineageFullGraphResponse(BaseModel):
    """Full lineage graph (all tables, optional database filter)."""

    nodes: list[LineageNodeResponse]
    edges: list[LineageEdgeResponse]

File: ai-data-dictionary/apps/frontend/src/lib/api/lineage.ts
/**
 * Lineage API: full graph (nodes + edges), optional database filter.
 */

import { apiClient } from "../api-client";
import { AuthStore } from "../auth-store";

export interface LineageNode {
  id: number;
  name: string;
  schema_name: string;
  table_type: string;
  level: number;
}

export interface LineageEdgeColumnPair {
  referenced_column?: string;
  referencing_column?: string;
}

export interface LineageEdge {
  source: number;
  target: number;
  relationship_type: string;
  label?: string | null;
  column_mapping?: LineageEdgeColumnPair[] | null;
  cardinality?: string | null; // one_to_one, one_to_many, many_to_many
  is_join_table?: boolean | null;
}

export interface LineageFullGraphResponse {
  nodes: LineageNode[];
  edges: LineageEdge[];
}

export interface LineageGraphResponse extends LineageFullGraphResponse {
  root_table_id: number;
}

function getToken(): string | null {
  return AuthStore.getToken();
}

export async function getFullLineage(databaseId?: number): Promise<LineageFullGraphResponse> {
  const qs = databaseId != null ? `?database_id=${databaseId}` : "";
  return apiClient.get<LineageFullGraphResponse>(`/lineage${qs}`, getToken() ?? undefined);
}

export async function getTableLineage(tableId: number): Promise<LineageGraphResponse> {
  return apiClient.get<LineageGraphResponse>(
    `/tables/${tableId}/lineage`,
    getToken() ?? undefined
  );
}
